doc/developer/design/20210715_table_update.md

```diff
@@ -1,52 +1,158 @@
+- Signals those waiting on the lock of its availability and lets one of them
+    proceed
+
+While the mutex I chose comes from `tokio::sync`, mutexes are well-known,
+well-tested, idiomatic tools from standard libraries of many programming
+languages, so using one has the advantage of giving us exactly what we want with
+very little work, and is generally less fallible than trying to hand-roll the
+same semantics using another implemenation.
+
+By using a mutex, if the session holding the write lock gets dropped, the lock
+is released. To reach parity with the `Option<u32>` approach, we would need to
+implement an "unlock" operation on `Drop`, at which point we're basically
+handrolling a mutex.
+
+#### Considerations
+
+**Implementation**
+
+I chose `tokio::sync::Mutex` because its
+[documentation](https://docs.rs/tokio/1.11.0/tokio/sync/struct.Mutex.html) states:
+
+> [`tokio::sync::Mutex`'s] lock guard is designed to be held across `.await`
+> points
+
+And we want to hold this lock across at least one `await` point during commits.
+There are more details in the documentation.
+
+While this has a potential performance implication, performance is expressly not
+a goal of this code.
+
+**Where to put the mutex handle**
+
+Because e.g. `ReadThenWrite` nodes are "long-lived" calls, the simplest place to
+share data (i.e. the mutex's handle) for the life of the function call is on the
+`Session`. However, because `Session` is even-longer-lived, this means that
+there is not a naturally occurring place for the mutex to release itself, so you
+must manually release the lock.
+
+While this isn't ideal, it isn't inferior to the prior iteration's design, which
+still required manual unlocking, but without the benefit of unlocking on drop.
+
+**Deferred work**
+
+This iteration of the code works similarly to the prior version, where deferred
+work is placed on a `VecDeque`. When an operation fails to obtain the lock, it's
+parked in a greenthread and uses `internal_cmd_tx` to signal the lock's
+availability.
+
+The previous iteration of this code removed the deferred work from the internal
+`VecDeque` and sent the entire plan through `internal_cmd_tx`. However, the
+mutex-backed implementation simply lets us send along the mutex's handle and
+lets the coordinator itself handle dequeuing the work and installing the handle
+on the appropriate session.
+
+There is also an alternate design that would be beneficial if we have multiple
+write locks (e.g. one per table), where installing the handles occur in the
+greenthread, i.e. wait on the mutex in the greenthread, install the handle on
+the session once it become available, and then message the entire plan to
+execute. However, this design is a little messier when dealing with
+cancellations, so I don't think is worth pursuing at this point.
+
+I believe the implemented approach also provides a linearizable guarantee: items
+are only ever enqueued in the order they're received, and dequeued unless
```

Comment:
I think that linearizability cares about the order of things relative to their completed order, not received order. It's still linearizable even if processed in a random order.

Reply:
Ah, ok! That makes sense. Will amend.